### ISSUE TITLE
Potential collapsible_if false positive

### ISSUE THREAD

**Issue by kbknapp** [Manishearth/rust-clippy#1489]
Clippy warns an if block could be collapsed, but the inner if contains computationally expensive `else if` blocks.

gist of the erroring code:

```rust

if foo {
    // ...
} else if bar {
    // ...
} else {
   if baz {
        // ...
   } else if qux {
        // ...
   } else if quux {
        // ...
   }
}
```

The problem is `qux` and `quux` are both computationally expensive and only potentially apply if `foo` and `bar` do not. Collapsing `baz` into an `else if`requires *also* collapsing `qux` and `quux` into the same `else if` chain and get evaluated even though there's no way for them to be `true`.

**Comment by llogiq** [comment]
I don't get it. `if a() { .. } else { if b() { .. } }` is *semantically equal* to `if a() { .. } else if b() { .. }` - `b` of course only gets evaluated if `a()` returns `false`. In your example any `true` `foo`, `bar` or `baz` will lead to the whole expression terminate without evaluating `qux` or even `quux` in both cases.

**Comment by kbknapp** [comment]
I'm a dummy - *hangs head in shame* 😄

**Comment by clarcharr** [comment]
Collapsing the if is equivalent; `quz` and `quux` are only evaluated if `baz` is false because of short-circuiting.

So this isn't a false positive.

**Comment by llogiq** [comment]
@kbknapp could have been me (especially this early in the morning – luckily I already had coffee :smile:).
